### PR TITLE
Twostream feature

### PR DESCRIPTION
**Pull request by marijnl**
This could be done more elegantly probably with a TwoStreamSampler but it would need some more time to implement.

#31

### CODE REVIEW

**Comment by KevinMusgrave** [comment]
This looks great! I'm a bit busy right now, but I'll try to take a look this weekend.

**Comment by KevinMusgrave** [comment]
Can you move your ipython notebook to a google colab notebook, like the other ones here: https://github.com/KevinMusgrave/pytorch-metric-learning/tree/master/examples

**Comment by marijnl** [comment]
I moved the notebook to colab. I tested it by installing the package from my branch

**Comment by preachermaan** [comment]
Awesome, thanks guys. I've been distracted with another project, but I plan to try and get this working over the next couple of weeks so I'll let you know if there's anything I can add.

**Comment by marijnl** [comment]
@KevinMusgrave don't worry take your time

**Comment by KevinMusgrave** [comment]
@marijnl I made some changes which you can find in [this commit](https://github.com/KevinMusgrave/pytorch-metric-learning/commit/d4005d17c60ea9b03a0c3e48a34fef9c0be46b96)

The main changes are:
- For the tester:
    - Only accept ```"compared_to_self"``` for the ```reference_set``` argument
    - Stack the anchor and posneg embeddings so that they can be used by the optional visualizer. Unstack them in ```set_reference_and_query```
- For the trainer:
    - Got rid of the "TwoStream" name requirement as I think it's too restrictive
    - Moved the ```subset_batch_miner``` exception functionality to the ```allowed_mining_funcs_keys``` function
    - Added ```lmu.get_all_triplets_indices(labels)``` as the default if ```tuple_miner``` isn't specified.

By the way, in your notebook I don't think ```minerIdentity``` is defined anywhere

**Comment by KevinMusgrave** [comment]
@marijnl Re: that minerIdentity in the notebook, don't worry about it, I'm creating a new notebook which is basically the same as yours, but follows the same format as the other notebooks.

**Comment by marijnl** [comment]
@kevinmusgrave great! Thanks a lot for reviewing this PR. I'm not yet very familiar with this package to see a lot of the negative trails. Thanks!

Something else maybe @preachermaan is also interested in. For my specific use case I also uses the notion of known very hard negatives. I mine negatives from Google similar image search and iteratively label them by hand. I look for negatives that are very similar to the user query but are not a positive. For now the twostream training is restricted to positive pairs which is somewhat to restrictive. I'm aiming to see if I can get this to work somehow

**Comment by preachermaan** [comment]
Oh right @marijnl. I'm confused. You mean you're not training with negatives right now? I thought most distance based metric learning approaches in some way relied on maximising the distance between examples of positive and negative class given an anchor. In my case the anchor will always be fixed i.e. a low quality phone based photo of a product, and I want to compare it to a fixed list of high quality stock photos of the actual products. In my scenario I'm looking for an exact match but I also have a dataset of known hard matches, i.e. where the stock photos are very similar. I haven't looked into the best approach yet, but I presume I'll need some kind of online triplet mining to get the best results.

**Comment by marijnl** [comment]
Sorry for the confusion. I do train with both positives and negatives. I seems our case is very similar. I have low quality phone images as well and their corresponding positive class. 

I extend my dataset regularly by "mining" visual similar images online. This means I lookup positives on Google images to get difficult negatives (I hand label to make sure it's not a positive). 

Now the problem with the current implementation  is that it only works on batches of tuples (anchor,positive). Since I don't have an anchor for visual similar images I cannot used them in this training procedure. 

In order to make this work I need to be able to add negatives as well

**Comment by KevinMusgrave** [comment]
@preachermaan The negatives are formed during training, based on labels.

**Comment by marijnl** [comment]
I think eventually I need to combine online and offline sampling

**Comment by preachermaan** [comment]
Okay, makes sense. From what I've read, triplet mining approach makes a huge difference. But like you I still see a scenerio where having fixed hard negatives would be useful - though probably prone to overfitting.
